### PR TITLE
Add information from doc/hacking.md and doc/make_cheatsheet.md back i…

### PR DESCRIPTION
**Pull request by jemmaissroff**
…nto contributing docs

[The PR](https://github.com/ruby/ruby/pull/5899) which refactored the contributing docs unintentionally excluded information from `doc/hacking.md` and `doc/make_cheatsheet.md`. This PR adds that information back into the docs, under the new cohesive structure.

### CODE REVIEW

File: doc/contributing/building_ruby.md [hunk 1]
```diff
@@ -72,3 +100,41 @@ with the Ruby script you'd like to run. You can use the following make targets:
...
+
+``` shell
+./autogen.sh
+mkdir build && cd build
+export ASAN_OPTIONS="halt_on_error=0:use_sigaltstack=0:detect_leaks=0"
+../configure cppflags="-fsanitize=address -fno-omit-frame-pointer" optflags=-O0 LDFLAGS="-fsanitize=address -fno-omit-frame-pointer"
+make
+```
+
+On Linux it is important to specify -O0 when debugging and this is especially true for ASAN which sometimes works incorrectly at higher optimisation levels.
```
Comment (st0012) [inline doc/contributing/building_ruby.md:116]:
```suggestion
On Linux it is important to specify `-O0` when debugging. This is especially true for ASAN which sometimes works incorrectly at higher optimisation levels.
```

File: doc/contributing/building_ruby.md [hunk 2]
```diff
@@ -72,3 +100,41 @@ with the Ruby script you'd like to run. You can use the following make targets:
...
+```
+
+If you need only C code coverage, you can remove `COVERAGE=true` from the above process.
+You can also use `gcov` command directly to get per-file coverage.
+
+If you need only Ruby code coverage, you can remove `--enable-gcov`.
+Note that `test-coverage.dat` accumulates all runs of `make test-all`.
+Make sure that you remove the file if you want to measure one test run.
+
+You can see the coverage result of CI: https://rubyci.org/coverage
```
Comment (nobu) [inline doc/contributing/building_ruby.md:140]: Missing newline at EOF.

File: doc/contributing/testing_ruby.md [hunk 3]
```diff
@@ -40,7 +47,13 @@ We can run any of the make scripts [in parallel](building_ruby.md#label-Running+
     make test-all
     ```
 
-    We can run a specific test file in this suite using the `TESTS` environment variable, for example:
+    We can run a specific test directory in this suite using the `TESTS` environment variable, for example:
```
Comment (nobu) [inline doc/contributing/testing_ruby.md:50]: Accurately, this is not an environment variable.

File: doc/contributing/building_ruby.md [hunk 4]
```diff
@@ -25,17 +25,27 @@
...
+
+    ``` shell
+    ../configure --disable-install-doc
+    ```
+
+6. [Run tests](testing_ruby.md) to confirm your build succeeded
+
+### Unexplainable Build Errors
+
+If you are having unexplainable build errors, after saving all your work, try running `git clean -xfd` in the source root to remove all git ignored local files. If you are working from a source directory that's been updated several times, you may have temporary build artefacts from previous releases which can cause build failures.
```
Comment (peterzhu2118) [inline doc/contributing/building_ruby.md:48]: Typo:

```suggestion
If you are having unexplainable build errors, after saving all your work, try running `git clean -xfd` in the source root to remove all git ignored local files. If you are working from a source directory that's been updated several times, you may have temporary build artifacts from previous releases which can cause build failures.
```

File: doc/contributing/building_ruby.md [hunk 5]
```diff
@@ -44,13 +54,31 @@ about Ruby's build to help out.
 
 ### Running make scripts in parallel
 
-To run make scripts in parallel, pass flag `-j<number of processes>`. For instance,
+To run a specific make script in parallel, pass flag `-j<number of processes>`. For instance,
 to run tests on 8 processes, use:
```
Comment (nobu) [inline doc/contributing/building_ruby.md:58]: This option is particular for some implementations, i.e., GNU make, BSD make.
It should be mentioned clearly.

File: doc/contributing/building_ruby.md [hunk 6]
```diff
@@ -25,17 +25,27 @@
...
     ./autogen.sh
-    mkdir build && cd build # its good practice to build outside of source dir
+    mkdir build && cd build # it's good practice to build outside of source dir
     mkdir ~/.rubies # we will install to .rubies/ruby-master in our home dir
     ../configure --prefix="${HOME}/.rubies/ruby-master"
     make install
     ```
 
-5. [Run tests](testing_ruby.md) to confirm your build succeeded
+5. Optional: If you are frequently building Ruby, disabling documentation will reduce the time it takes to `make install`:
```
Comment (ioquatix) [inline doc/contributing/building_ruby.md:38]:
```suggestion
5. Optional: If you are frequently building Ruby, disabling documentation will reduce the time it takes to `make`:
```

File: doc/contributing/testing_ruby.md (hunk 3 above)
Comment (ioquatix) [reply doc/contributing/testing_ruby.md:50]:
```suggestion
    We can run a specific test directory in this suite using the `TESTS` option, for example:
```

File: doc/contributing/testing_ruby.md [hunk 7]
```diff
@@ -40,7 +47,13 @@ We can run any of the make scripts [in parallel](building_ruby.md#label-Running+
...
+
+    ```
+    make test-all TESTS=test/rubygems
+    ```
+
+    We can run a specific test file in this suite by also using the `TESTS` environment variable, for example:
```
Comment (ioquatix) [inline doc/contributing/testing_ruby.md:56]:
```suggestion
    We can run a specific test file in this suite by also using the `TESTS` option, for example:
```

**Review by ioquatix — APPROVED** [review]
This looks really good to me.

If we don't already have it, advice around running testing (from btest through to test-spec) might be useful, as in "here are the tests you should run locally before submitting a PR".

File: doc/contributing/building_ruby.md (hunk 4 above)
Comment (jemmaissroff) [reply doc/contributing/building_ruby.md:48]: I don't mind changing this, but it's a case of [British vs American spelling](https://grammarist.com/spelling/artefact-artifact/). Do Ruby docs generally adhere to American spellings over British ones?

**Comment by jemmaissroff** [comment]
> If we don't already have it, advice around running testing (from btest through to test-spec) might be useful, as in "here are the tests you should run locally before submitting a PR".

Thanks for the review @ioquatix. Adding [the advice around running testing](https://docs.ruby-lang.org/en/master/contributing/testing_ruby_md.html) was part of [the previous PR](https://github.com/ruby/ruby/pull/5899) and is linked from [the main contributing page](https://docs.ruby-lang.org/en/master/contributing_md.html)!

File: doc/contributing/building_ruby.md (hunk 4 above)
Comment (peterzhu2118) [reply doc/contributing/building_ruby.md]: TIL "artefact" is British spelling. I think the docs defaults to American English.

**Comment by ioquatix** [comment]
Great, that's really awesome, thanks for the clarification.

**Review by peterzhu2118 — APPROVED** [review]
Thank you for working on this!

**Pull request merged**